writer.go

```diff
@@ -308,50 +308,83 @@ func (w *Writer) WriteMessages(ctx context.Context, msgs ...Message) error {
 		res = make(chan error, len(msgs))
 	}
 	t0 := time.Now()
-	defer w.stats.writeTime.observeDuration(time.Since(t0))
 
-	w.mutex.RLock()
-	closed := w.closed
-	w.mutex.RUnlock()
+	for attempt := 0; attempt < w.config.MaxAttempts; attempt++ {
+		w.mutex.RLock()
 
-	if closed {
-		return io.ErrClosedPipe
-	}
-
-	for i, msg := range msgs {
+		if w.closed {
+			w.mutex.RUnlock()
+			return io.ErrClosedPipe
+		}
 
-		if int(msg.size()) > w.config.BatchBytes {
-			err := MessageTooLargeError{
-				Message:   msg,
-				Remaining: msgs[i+1:],
+		for i, msg := range msgs {
+			if int(msg.size()) > w.config.BatchBytes {
+				err := MessageTooLargeError{
+					Message:   msg,
+					Remaining: msgs[i+1:],
+				}
+				w.mutex.RUnlock()
+				return err
+			}
+			select {
+			case w.msgs <- writerMessage{
+				msg: msg,
+				res: res,
+			}:
+			case <-ctx.Done():
+				w.mutex.RUnlock()
+				return ctx.Err()
 			}
-			return err
 		}
 
-		wm := writerMessage{msg: msg, res: res}
+		w.mutex.RUnlock()
```

Comment:
Note: I don't understand why we keep this mutex for so long in this version of the code. We only seem to need to lock `w.closed`